### ISSUE TITLE
cicd-wait not working like I would expect.

### ISSUE THREAD

**Issue by ForceConstant** [TobiWo/eth-duties#43]
From reading the Readme, and my ideas of what I need. I figured "--mode cicd-wait" would have the executable run and then exit after an attestation. If this is what it does, it is not working for me, and it seems to just keep going until the 780 second timeout. Do I understand wrong? If so, how can I get this functionality?

**Comment by TobiWo** [comment]
Hi @ForceConstant thanks for reaching out.

The mode `cicd-wait` waits for a certain amount of time until there is no duty in the pipeline. If you have scanning for attestation duties activated (which is default) it will wait the defined amount of time (780 secs.) until it exits not gracefully. This is because there is a constant inflow of attestation duties so the queue will never be empty and the graceful exit is only performed when no duty is in the queue. So with activated attestation duties this cannot be reached. The idea behind the mode is to use it in automated deployment pipelines and probably concentrate on 'important' duties like sync-committee or block proposals so that you do not update your infra when one of these duties is in the queue.
So to summarize `eth-duties` does not support what you need currently. I would need to add this functionality. Anyway, I see your point. May I ask you what you try to achieve in waiting until one attestation is done and then do the exit?

**Comment by ForceConstant** [comment]
Use case would be to automate certain tasks to maximize time between attestation, such as needing to update clients, reboot node, etc.

**Comment by TobiWo** [comment]
Ok, thanks. I think this sounds like an interesting extension for the cicd modes for home stakers (for optimization purposes). My idea would be the following:

- I will most probably modify the current `cicd-wait`
- `eth-duties` will exit gracefully if the next attestation is longer away than a defined period of time e.g. 7 mins.
- if the next attestation is closer e.g. 4 mins. the application will keep on running in the wait mode but exit not gracefully in the `cicd-exit` mode
- the time threshold can bet set by the user ofc
- all this will be only relevant if no sync committee duty or block proposal is in the queue

What do you think?

**Comment by ForceConstant** [comment]
Sounds good.

**Comment by TobiWo** [comment]
Hi @ForceConstant, sorry for the development delay. Couldn't work on `eth-duties` due to private stuff. However, I will be able to work on the tool more constantly from now on. In best case I will finish this feature tomorrow. Lookout for release `v0.4.0`.

**Comment by TobiWo** [comment]
Hi @ForceConstant , I think I'm done with the new feature. It would be super helpful if you can test it out before I create a new release. Do you use a docker setup? If yes you can test the new feature with the following image `docker pull tobiwo/eth-duties:dev-cicd`. If you use the binary you could build it on your own but this is a bit more tricky if you are not familiar with poetry and the whole python workflow.

You can find the updated documentation [here](https://github.com/TobiWo/eth-duties/tree/feature/improve-cicd-modes-for-home-stakers#eth-duties). Any feedback on whether the documentation is understandable or where it can be improved is highly appreciated.

**Comment by ForceConstant** [comment]
Thanks, so I tried it out, and it actually works better than I had thought. I like the fact that it doesn't just exit when the last attestation goes out, but only if the next one is then a min of time away. I tested with '--mode cicd-wait --mode-cicd-attestation-time 240 --mode-cicd-attestation-proportion 0.8' and it worked great. Thanks.

**Comment by TobiWo** [comment]
I'm happy that you like it. I think it is a super nice feature for home stakers so thanks again for your suggestions :smiley: 
BTW: Compared to yesterday I added more verbosity so that you e.g. see with which exit code the application exits or which proportion of your monitored attestation duties will be executed in your defined time threshold or later. I updated the dev image if you want to test again.

The PR will be reviewed in the next days hopefully. Please be aware that there might be smaller changes in the final release if the reviewer finds something.
